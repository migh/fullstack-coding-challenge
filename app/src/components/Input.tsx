import * as React from 'react';

export interface InputProps {
  onChange: (event: React.FormEvent<HTMLInputElement>) => void;
  onEnter: (event: React.FormEvent<HTMLInputElement>) => void;
  placeholder: string;
  value: string;
}

export const Input = (props: InputProps) => (
    <input
      className="form-control"
      type="text"
      placeholder={props.placeholder}
      value={props.value}
      onChange={props.onChange}
      onKeyPress={props.onEnter}
    />
);

