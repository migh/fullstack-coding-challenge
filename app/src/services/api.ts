export interface Options {
  url?: string;
  type?: string;
  init?: any;
}

/**
 * Basic API request function
 * As the project grows and depending on the architecture
 * it could make sense to put this into a class.
 * It is promise based but at this point very general.
 * It is hard to test internal function properly.
 */
export function request(options: Options) {
  const { url, type, init } = options;


  function requestResolver(resolve: () => any, reject: () => any) {
  fetch(url, init ? {...init} : {})
      .then((response: any) => response[type]())
      .then(resolve)
      .catch(reject);
  }

  return new Promise(requestResolver);
}

/**
 * Hits /translations endpoint
 * Gets all current translations
 */
export function getTranslations() {
  return request({
    url: '/translations',
    type: 'json'
  });
}

/**
 * Hits /translation/:id endpoint
 * Gets translation status, id is job id in unbabel api
 */
export function checkStatus(jobId: string) {
  return request({
    url: `/translation/${jobId}`,
    type: 'json'
  });
}

/**
 * Hits /translate endpoint
 * Request a translation
 */
export function translate(text: string) {
  return request({
    url: '/translate',
    type: 'text',
    init: {
      method: 'POST',
      body: JSON.stringify({ text }),
      headers: {
        'Content-Type': 'application/json'
      }
    }
  });
}
