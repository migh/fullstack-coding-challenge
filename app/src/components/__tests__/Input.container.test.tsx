import * as React from 'react';
import { shallow } from 'enzyme';

import * as api from '../../services/api';
import { InputContainer } from '../Input.container';

jest.mock('../../services/api', () => ({
  getTranslations: jest.fn(() => Promise.resolve([])),
  translate: jest.fn(() => Promise.resolve({})),
  checkStatus: jest.fn(() => Promise.resolve({}))
}));

describe('<InputContainer />', () => {
  describe('Snapshots', () => {
    it('should render.', () => {
      expect(shallow(<InputContainer />)).toMatchSnapshot();
    });
  });

  describe('updateTranslations()', () => {
    it('should call api.getTranslations.', () => {
      const wrapper = shallow(<InputContainer />);
      const instance = (wrapper.instance() as InputContainer);

      instance.updateTranslations();

      expect(api.getTranslations).toHaveBeenCalled();
    });
  });

  describe('translateClickHandler()', () => {
    it('should call api.translate.', () => {
      const wrapper = shallow(<InputContainer />);
      const instance = (wrapper.instance() as InputContainer);

      instance.translateClickHandler();

      expect(api.translate).toHaveBeenCalled();
    });
  });
});
