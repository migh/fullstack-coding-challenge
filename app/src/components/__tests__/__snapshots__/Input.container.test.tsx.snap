// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> Snapshots should render. 1`] = `ShallowWrapper {}`;
